docs/reference/setup/bootstrap-checks.asciidoc

```diff
@@ -0,0 +1,144 @@
+[[bootstrap-checks]] == Bootstrap Checks
+
+Collectively, we have a lot of experience with users suffering
+unexpected issues because they have not configured
+<<important-setting,important settings>>. In previous versions of
+Elasticsearch, some of these settings were logged as
+warnings. Understandably, users sometimes miss these log messages. To
+ensure that these settings receive the attention that they deserve,
+Elasticsearch has bootstrap checks upon startup.
+
+These bootstrap checks inspect a variety of Elasticsearch and system
+settings and compare them to values that are safe for the operation of
+Elasticsearch. If Elasticsearch is in development mode, any bootstrap
+checks that fail appear as warnings in the Elasticsearch log. If
+Elasticsearch is in production mode, any bootstrap checks that fail will
+cause Elasticsearch to refuse to start.
+
+=== Development vs. production mode
+
+By default, Elasticsearch binds and publishes to `localhost`. This is
+fine for downloading and playing with Elasticsearch, and everyday
+development but it's useless for production systems. For a production
+installation to be reachable, it must either bind or publish to an
+external interface. Thus, we consider Elasticsearch to be in development
+mode if it does not bind nor publish to an external interface (the
+default), and is otherwise in production mode if it does bind or publish
+to an external interface.
+
+=== Heap size check
+
+If a JVM is started with unequal initial and max heap size, it can be
+prone to pauses as the JVM heap is resized during system usage. To avoid
+these resize pauses, it's best to start the JVM with the initial heap
+size equal to the maximum heap size. Additionally, if
+<<[[bootstrap.mlockall]] is enabled, the JVM will lock the initial size
+of the heap on startup. If the initial heap size is not equal to the
+maximum heap size, after a resize it will not be the case that all of
+the JVM heap is locked in memory. To pass the heap size check, you must
+configure the [[heap-size,heap size]].
+
+=== File descriptor check
+
+File descriptors are a Unix construct for tracking open "files".
+"Files" though could be a physical file, a virtual file (e.g.,
+``/proc/loadavg`), or network sockets. Elasticsearch requires lots file
+descriptors (e.g., every shard is composed of multiple segments and
+other files, plus connections to other nodes, etc.). This bootstrap
+check is enforced on OS X and Linux. To pass the file descriptor check,
+you might have to configure [[file-descriptors,file descriptors]].
+
+=== Memory lock check
+
+A terrible thing can happen if parts of the JVM heap are swapped out to
```

Comment:
I might reword "When the JVM does a major garbage collection it touches every page of the heap. If any of those pages are swapped out to disk they will have to be swapped back in to memory. That causes lots of disk activity that Elasticsearch would much rather be used to service request...."